src/PackagePrivate/DatabaseTermIdsResolver.php

```diff
@@ -0,0 +1,129 @@
+<?php
+
+namespace Wikibase\TermStore\MediaWiki\PackagePrivate;
+
+use InvalidArgumentException;
+use stdClass;
+use Wikimedia\Rdbms\IDatabase;
+use Wikimedia\Rdbms\ILoadBalancer;
+use Wikimedia\Rdbms\IResultWrapper;
+
+/**
+ * Term ID resolver using the normalized database schema.
+ *
+ * @license GPL-2.0-or-later
+ */
+class DatabaseTermIdsResolver implements TermIdsResolver {
+
+	/** @var TypeIdsResolver */
+	private $typeIdsResolver;
+
+	/** @var ILoadBalancer */
+	private $lb;
+
+	/** @var IDatabase */
+	private $dbr = null;
+
+	/** @var IDatabase */
+	private $dbw = null;
+
+	/** @var string[] stash of data returned from the {@link TypeIdsResolver} */
+	private $typeNames = [];
+
+	public function __construct(
+		TypeIdsResolver $typeIdsResolver,
+		ILoadBalancer $lb
+	) {
+		$this->typeIdsResolver = $typeIdsResolver;
+		$this->lb = $lb;
+	}
+
+	/*
+	 * Term data is first read from the replica; if that returns less rows than we asked for,
+	 * then there are some new rows in the master that were not yet replicated, and we fall back
+	 * to the master. As the internal relations of the term store never change (for example,
+	 * a term_in_lang row will never suddenly point to a different text_in_lang), a master fallback
+	 * should never be necessary in any other case. However, callers need to consider where they
+	 * got the list of term IDs they pass into this method from: if it’s from a replica, they may
+	 * still see outdated data overall.
+	 */
+	public function resolveTermIds( array $termIds ): array {
+		$terms = [];
+		$this->connectDbr();
+
+		$replicaResult = $this->selectTerms( $this->dbr, $termIds );
+		$this->preloadTypes( $replicaResult );
+		$replicaTermIds = [];
+
+		foreach ( $replicaResult as $row ) {
+			$replicaTermIds[] = $row->wbtl_id;
+			$this->addResultTerms( $terms, $row );
+		}
+
+		if ( count( $replicaTermIds ) !== count( $termIds ) ) {
+			$masterTermIds = array_values( array_diff( $termIds, $replicaTermIds ) );
+			$this->connectDbw();
+			$masterResult = $this->selectTerms( $this->dbw, $masterTermIds );
+			$this->preloadTypes( $masterResult );
```

Comment:
the wrapping logic here, achieved by calls to `preloadTypes( $masterResult )` and `lookupType` around calls to `addResultTerms`, is not easy to read than if it was implemented in a pure fashion.

PR #34 against this one suggest how to purify `addResultTerms`.